File: app/src/main/java/ch/unstable/ost/LastQueryCardFragment.java
package ch.unstable.ost;


import android.content.Context;
import android.content.Intent;
import android.os.Bundle;
import android.support.annotation.MainThread;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Button;
import android.widget.TextView;

import com.google.common.base.Preconditions;

import ch.unstable.ost.api.model.ConnectionQuery;
import ch.unstable.ost.database.Databases;
import ch.unstable.ost.database.dao.QueryHistoryDao;
import ch.unstable.ost.database.model.QueryHistory;
import io.reactivex.android.schedulers.AndroidSchedulers;
import io.reactivex.disposables.CompositeDisposable;
import io.reactivex.disposables.Disposable;
import io.reactivex.schedulers.Schedulers;

import static ch.unstable.ost.views.lists.query.QueryBinderKt.bindQuery;

public class LastQueryCardFragment extends QuickstartCardFragment {
    private static final String TAG = "LastQueryCardFragment";
    private View mCardLastQuery;
    private QueryHistoryDao mQueryDao;
    private TextView mLastQueryFromTo;
    private TextView mLastQueryDate;
    private OnQuerySelectedListener mOnQuerySelectedListener;
    private QueryHistory mLastQuery;
    private CompositeDisposable mCompositeDisposable;

    @Override
    public void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        if (mQueryDao == null) {
            mQueryDao = Databases.getCacheDatabase(getContext()).queryHistoryDao();
        }
        mCompositeDisposable = new CompositeDisposable();
    }

    @Override
    public void onAttach(Context context) {
        super.onAttach(context);
        Preconditions.checkState(
                context instanceof OnQuerySelectedListener,
                "context must implement OnQuerySelectedListener");
        mOnQuerySelectedListener = (OnQuerySelectedListener) context;
    }


    @Override
    public void onDetach() {
        super.onDetach();
        mOnQuerySelectedListener = null;
    }

    @Nullable
    @Override
    public View onCreateView(LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        return inflater.inflate(R.layout.card_last_query, container, false);
    }

    @Override
    public void onViewCreated(View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
        mCardLastQuery = view.findViewById(R.id.cardLastQuery);
        mCardLastQuery.setVisibility(View.GONE);
        mLastQueryFromTo = view.findViewById(R.id.lastQueryFromTo);
        mLastQueryDate = view.findViewById(R.id.lastQueryDate);

        Button buttonMore = view.findViewById(R.id.buttonMore);
        buttonMore.setOnClickListener(eventView -> onShowMoreQueries());

        Button buttonLoad = view.findViewById(R.id.buttonOpen);
        buttonLoad.setOnClickListener(eventView -> onOpenQuery());
    }

    private void onOpenQuery() {
        if (mOnQuerySelectedListener == null) {
            Log.w(TAG, "mOnQuerySelectedListener is null");
            return;
        }
        //noinspection ResultOfMethodCallIgnored
        Preconditions.checkNotNull(mLastQuery, "mLastQuery is null");
        mOnQuerySelectedListener.onRouteSelected(mLastQuery.getQuery());
    }

    private void onShowMoreQueries() {
        Intent intent = new Intent(getContext(), QueryHistoryActivity.class);
        startActivity(intent);
    }

    @Override
    public void onResume() {
        super.onResume();
        Disposable disposable = mQueryDao.getLatestQuery()
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(this::updateLatestQuery, getErrorConsumer());
        mCompositeDisposable.add(disposable);
    }

    @Override
    public void onDestroy() {
        super.onDestroy();
        mCompositeDisposable.dispose();
    }

    @MainThread
    private void updateLatestQuery(QueryHistory queryHistory) {
        mCardLastQuery.setVisibility(View.VISIBLE); // TODO animate
        mLastQuery = queryHistory;
        bindQuery(queryHistory, mLastQueryDate, mLastQueryFromTo);
    }

    @Override
    protected int getErrorMessage() {
        return R.string.error_failed_to_load_last_query;
    }

    public interface OnQuerySelectedListener {
        @MainThread
        void onRouteSelected(@NonNull ConnectionQuery query);
    }
}

File: app/src/main/java/ch/unstable/ost/LastQueryCardFragment.kt
package ch.unstable.ost

import android.content.Context
import android.content.Intent
import android.os.Bundle
import android.util.Log
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.annotation.MainThread
import ch.unstable.ost.api.model.ConnectionQuery
import ch.unstable.ost.database.Databases
import ch.unstable.ost.database.dao.QueryHistoryDao
import ch.unstable.ost.database.model.QueryHistory
import ch.unstable.ost.databinding.CardLastQueryBinding
import ch.unstable.ost.views.lists.query.QueryBinder.bindQuery
import io.reactivex.android.schedulers.AndroidSchedulers
import io.reactivex.disposables.CompositeDisposable
import io.reactivex.functions.Consumer
import io.reactivex.schedulers.Schedulers

class LastQueryCardFragment : QuickstartCardFragment() {
    private var binding: CardLastQueryBinding? = null
    private lateinit var queryDao: QueryHistoryDao
    private var onQuerySelectedListener: OnQuerySelectedListener? = null
    private var lastQuery: QueryHistory? = null
    private var compositeDisposable: CompositeDisposable? = null

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        queryDao = Databases.getCacheDatabase(requireContext()).queryHistoryDao()
        compositeDisposable = CompositeDisposable()
    }

    override fun onDestroyView() {
        super.onDestroyView()
        this.binding = null
    }

    override fun onAttach(context: Context) {
        super.onAttach(context)
        onQuerySelectedListener = context as OnQuerySelectedListener
    }

    override fun onDetach() {
        super.onDetach()
        onQuerySelectedListener = null
    }

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View {
        val binding = CardLastQueryBinding.inflate(inflater, container, false)
        this.binding = binding
        binding.cardLastQuery.visibility = View.GONE
        binding.buttonMore.setOnClickListener { onShowMoreQueries() }
        binding.buttonOpen.setOnClickListener { onOpenQuery() }
        return binding.root

    }

    private fun onOpenQuery() {
        if (onQuerySelectedListener == null) {
            Log.w(TAG, "mOnQuerySelectedListener is null")
            return
        }
        lastQuery ?: error("mLastQuery is null")
        onQuerySelectedListener!!.onRouteSelected(lastQuery!!.query)
    }

    private fun onShowMoreQueries() {
        val intent = Intent(context, QueryHistoryActivity::class.java)
        startActivity(intent)
    }

    override fun onResume() {
        super.onResume()
        val disposable = queryDao.latestQuery
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(Consumer { queryHistory: QueryHistory -> updateLatestQuery(queryHistory) }, errorConsumer)
        compositeDisposable?.add(disposable)
    }

    override fun onDestroy() {
        super.onDestroy()
        compositeDisposable?.dispose()
    }

    @MainThread
    private fun updateLatestQuery(queryHistory: QueryHistory) {
        lastQuery = queryHistory
        val binding = this.binding
        if (binding != null) {
            binding.cardLastQuery.visibility = View.VISIBLE // TODO animate
            bindQuery(queryHistory, binding.lastQueryDate, binding.lastQueryFromTo)
        }
    }

    override val errorMessage: Int
        get() = R.string.error_failed_to_load_last_query

    interface OnQuerySelectedListener {
        @MainThread
        fun onRouteSelected(query: ConnectionQuery)
    }

    companion object {
        private const val TAG = "LastQueryCardFragment"
    }
}